crates/wasmtime/src/runtime/vm/provenance.rs

```diff
@@ -0,0 +1,215 @@
+//! code. Wasmtime has a nontrivial amount of `unsafe` code and when/where
+//! pointers are valid is something we need to be particularly careful about.
+//! All safe Rust does not need to worry about this module and only the unsafe
+//! runtime bits need to worry about it.
+//!
+//! In general Wasmtime does not work with Rust's strict pointer provenance
+//! rules. The primary reason for this is that Cranelift does not have the
+//! concept of a pointer type meaning that backends cannot know what values are
+//! pointers and what aren't. This isn't a huge issue for ISAs like x64 but for
+//! an ISA like Pulley Bytecode it means that the Pulley interpreter cannot
+//! respect strict provenance.
+//!
+//! > **Aside**: an example of how Pulley can't respect pointer provenance is
+//! > consider a wasm load. The wasm load will add a wasm address to the base
+//! > address of the host. In this situation what actually needs to happen is
+//! > that the base address of the host is a pointer which is byte-offset'd by
+//! > the wasm address. Cranelift IR has no knowledge of which value is
+//! > the wasm address and which is the host address. This means that Cranelift
+//! > can freely commute the operands of the addition. This means that when
+//! > executing Pulley doesn't know which values are addresses and which aren't.
+//!
+//! This isn't the end of the world for Wasmtime, however, it just means that
+//! when we run in MIRI we are restricted to "permissive provenance" or "exposed
+//! provenance". The tl;dr; of exposed provenance is that at certain points we
+//! declare a pointer as "this is now exposed". That converts a pointer to the
+//! `usize` address and then semantically (just for rustc/llvm mostly) indicates
+//! that the provenance of the pointer is added to a global list of provenances.
+//! Later on Wasmtime will execute an operation to convert a `usize` back into a
+//! pointer which will pick "the most appropriate provenance" from said global
+//! list of provenances.
+//!
+//! In practice we expect that at runtime all of these provenance-related ops
+//! are noops and compile away to nothing. The only practical effect that's
+//! expected is that some optimizations may be hindered in LLVM occasionally or
+//! something like that which is by-and-large what we want to happen. Note that
+//! another practical consequence of not working with "strict provenance" means
+//! that Wasmtime is incompatible with platforms such as CHERI where exposed
+//! provenance is not available.
+
+use crate::vm::SendSyncPtr;
+use core::fmt;
+use core::ptr::NonNull;
+use core::sync::atomic::{AtomicU64, AtomicUsize};
+use wasmtime_environ::VMSharedTypeIndex;
+
+/// A pointer that is used by compiled code, or in other words is accessed
+/// outside of Rust.
+///
+/// This is intended to be the fundamental data type used to share
+/// pointers-to-things with Cranelift-compiled code for example. An example of
```

Comment:
In these comments there are a lot of references to Cranelift, but it could also be Winch. Might be worth rewording "Cranelift-compiled code" to "compiled Wasm code" or something.